tapchannel/aux_traffic_shaper.go

```diff
@@ -0,0 +1,242 @@
+		return 0, fmt.Errorf("error decoding HTLC blob: %w", err)
+	}
+
+	localBalance := cmsg.OutputSum(commitment.LocalOutputs())
+
+	// There either already is an amount set in the HTLC (which would
+	// indicate it to be a keysend payment), in which case we don't need an
+	// RFQ ID as we can just compare the local balance and the required HTLC
+	// amount. If there is no amount set, we need to look up the RFQ ID in
+	// the HTLC blob and use the accepted quote to determine the amount.
+	htlcAssetAmount := htlc.Amounts.Val.Sum()
+	if htlcAssetAmount != 0 && htlcAssetAmount <= localBalance {
+		// We signal "infinite" bandwidth by returning a very high
+		// value (number of Satoshis ever in existence).
+		return lnwire.NewMSatFromSatoshis(btcutil.MaxSatoshi), nil
+	}
+
+	// If the HTLC doesn't have an RFQ ID, we can't use any channel.
+	if htlc.RfqID.ValOpt().IsNone() {
+		return 0, nil
+	}
+
+	rfqID := htlc.RfqID.ValOpt().UnsafeFromSome()
+	acceptedQuotes := s.cfg.RfqManager.PeerAcceptedSellQuotes()
+	quote, ok := acceptedQuotes[rfqID.Scid()]
+	if !ok {
+		return 0, fmt.Errorf("no accepted quote found for RFQ ID "+
+			"%x (SCID %d)", rfqID[:], rfqID.Scid())
+	}
+
+	mSatPerAssetUnit := quote.BidPrice
+
+	// The available balance is the local asset unit expressed in
+	// milli-satoshis.
+	return lnwire.MilliSatoshi(localBalance) * mSatPerAssetUnit, nil
+}
+
+// ProduceHtlcExtraData is a function that, based on the previous extra data
+// blob of an HTLC, may produce a different blob or modify the amount of bitcoin
+// this htlc should carry.
+func (s *AuxTrafficShaper) ProduceHtlcExtraData(totalAmount lnwire.MilliSatoshi,
+	htlcCustomRecords lnwire.CustomRecords) (lnwire.MilliSatoshi, tlv.Blob,
+	error) {
+
+	if len(htlcCustomRecords) == 0 {
+		return totalAmount, nil, nil
+	}
+
+	// We need to do a round trip to convert the custom records to a blob
+	// that we can then parse into the correct struct again.
+	htlcBlob, err := htlcCustomRecords.Serialize()
+	if err != nil {
+		return 0, nil, fmt.Errorf("error serializing HTLC blob: %w",
+			err)
+	}
+
+	htlc, err := rfqmsg.DecodeHtlc(htlcBlob)
+	if err != nil {
+		return 0, nil, fmt.Errorf("error decoding HTLC blob: %w", err)
+	}
+
+	// If we already have an asset amount in the HTLC, we assume this is a
+	// keysend payment and don't need to do anything. We even return the
+	// original on-chain amount as we don't want to change it.
+	if htlc.Amounts.Val.Sum() > 0 {
+		return totalAmount, htlcBlob, nil
+	}
+
+	if htlc.RfqID.ValOpt().IsNone() {
+		return 0, nil, fmt.Errorf("no RFQ ID present in HTLC blob")
+	}
+
+	rfqID := htlc.RfqID.ValOpt().UnsafeFromSome()
+	acceptedQuotes := s.cfg.RfqManager.PeerAcceptedSellQuotes()
+	quote, ok := acceptedQuotes[rfqID.Scid()]
+	if !ok {
+		return 0, nil, fmt.Errorf("no accepted quote found for RFQ ID "+
+			"%x (SCID %d)", rfqID[:], rfqID.Scid())
+	}
+
+	mSatPerAssetUnit := quote.BidPrice
+
+	numAssetUnits := uint64(totalAmount*10/mSatPerAssetUnit) / 10
```

Comment:
I think here we're converting the HTLC BTC amount to tap asset amount using the agreed RFQ quote. Is that right? I think we should just add a comment to explain that.

Reply:
Yeah, you're right. Added a comment.